### PR TITLE
Activities

### PR DESCRIPTION
**Pull request by apmcleod**
Separates the menu and the game into 2 different activities, and removes unused code and layout clutter (cpus, online game, animation). Partway through #17 , but I still want to look at separating the direction chooser as well.

### CODE REVIEW

File: src/com/tictacgo/TicTacGoGame.java [hunk 1]
```diff
@@ -0,0 +1,402 @@
+package com.tictacgo;
```
Comment (hopskipnfall) [inline src/com/tictacgo/TicTacGoGame.java]: Are there any changes to this file other than a rename?

File: src/com/tictacgo/TicTacGoGame.java [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+import android.widget.TableRow;
+import android.widget.TextView;
+
+import com.tictacgo.data.Board;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class TicTacGoGame extends Activity {
```
Comment (jimmyfnugent) [inline src/com/tictacgo/TicTacGoGame.java]: Rename this TicTacGoGameActivity (and same for Menu).

File: res/layout/game.xml [hunk 3]
```diff
@@ -27,11 +27,6 @@
     			android:layout_width="wrap_content"
 	    		android:layout_height="wrap_content"
 	    		android:text="New Game"/>
-    		<Button
```
Comment (jimmyfnugent) [inline res/layout/game.xml:30]: GitHub won't let me do file-level comments.

Canonically the layout files for activities are called activity_my_activity.xml.

File: src/com/tictacgo/TicTacGoGame.java [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+  /**
+   * The index of the current undo/redo history
+   */
+  private int historyIndex;
+
+  /**
+   * The initial turn selection. Used for the New Game Button
+   */
+  private Board.Player turn;
```
Comment (jimmyfnugent) [inline src/com/tictacgo/TicTacGoGame.java]: Might want to import Board.Player separately so you can just call these Players.

File: src/com/tictacgo/TicTacGoMenu.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  public void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.menu);
+
+    findViewById(R.id.playButton).setOnClickListener(new OnClickListener() {
+      public void onClick(View v) {
+        Intent playGame = new Intent(v.getContext(), TicTacGoGame.class);
+
+        String p1Name = LayoutUtils.getTextOrHint((TextView) findViewById(R.id.localPlayerOneName));
+        playGame.putExtra("p1Name", p1Name);
```
Comment (jimmyfnugent) [inline src/com/tictacgo/TicTacGoMenu.java]: This feels sloppy. Let's do 2 things:
1. Pull out `p1Name`, etc. into constants. Preferably, prefix them with "com.tictacgo" to prevent name collisions with intents from other apps.
2. Pass the starting Player as an extra instead of a resource ID (line 31).

File: src/com/tictacgo/TicTacGoGame.java [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+    if (first == R.id.localTurnSelectX) {
+      turn = Board.Player.X;
+    } else if (first == R.id.localTurnSelectO) {
+      turn = Board.Player.O;
+    } else {
+      turn = null;
+    }
+
+    // Undo/redo initialization
+    undoHistory = new ArrayList<>();
```
Comment (jimmyfnugent) [inline src/com/tictacgo/TicTacGoGame.java]: Sorry to flip-flop on this, but I'm actually not a huge fan of using the diamond operator where the type isn't declared immediately before it -- just reading this line, it's impossible for me to figure out what undoHistory is. Thoughts?

File: src/com/tictacgo/TicTacGoGame.java [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+    height = intent.getIntExtra("height", 300);
+    board = new Board(turn, height, getBaseContext());
+
+    // Set up the screen
+    ((TextView) findViewById(R.id.gamePlayerOneName)).setText(intent.getStringExtra("p1Name"));
+    ((TextView) findViewById(R.id.gamePlayerTwoName)).setText(intent.getStringExtra("p2Name"));
+    fl = (FrameLayout) findViewById(R.id.gameBoard);
+
+    // What to do when an empty space is clicked
+    pieceClicked = new View.OnClickListener() {
```
Comment (jimmyfnugent) [inline src/com/tictacgo/TicTacGoGame.java]: Can you put a TODO here to pull out the listeners into inner classes?

File: src/com/tictacgo/data/Piece.java [hunk 8]
```diff
@@ -149,6 +140,14 @@ public int getXPosition() {
 		return position[0];
 	}
 
+  public int getRow() {
+    return position[0] + 1;
```
Comment (jimmyfnugent) [inline src/com/tictacgo/data/Piece.java:148]: This is going to get us into trouble someday...

It actually already did for me once. We should consider using {0, 1, 2} everywhere.

File: src/com/tictacgo/data/Piece.java [hunk 9]
```diff
@@ -206,7 +205,7 @@ public Player getPlayer() {
 
     public void setPlayer(Player player) {
         this.player = player;
-        updateImageResource();
+        updateImageResource(true);
```
Comment (jimmyfnugent) [inline src/com/tictacgo/data/Piece.java:212]: I think this has some unmerged changes from other PRs? This was updated to be two separate updateImageResource methods.

File: src/com/tictacgo/TicTacGoGame.java (hunk 2 above)
Comment (apmcleod) [inline src/com/tictacgo/TicTacGoGame.java]: Ok, renamed them.

File: res/layout/game.xml (hunk 3 above)
Comment (apmcleod) [inline res/layout/game.xml:30]: Alright, renamed the 2 activity layout xml files.

File: src/com/tictacgo/TicTacGoGame.java (hunk 4 above)
Comment (apmcleod) [inline src/com/tictacgo/TicTacGoGame.java]: Done.

File: src/com/tictacgo/TicTacGoMenu.java (hunk 5 above)
Comment (apmcleod) [inline src/com/tictacgo/TicTacGoMenu.java]: 1. Yes, I'll add those as constants.
2. Makes sense. I believe this will involve having the enum implement Parcelable.

EDIT: Ok, 1 and 2 are both done. I didn't know enums were Serializable by default, so this was easy.

File: src/com/tictacgo/TicTacGoGame.java (hunk 6 above)
Comment (apmcleod) [inline src/com/tictacgo/TicTacGoGame.java]: I could go either way. I just don't like yellow bars on the files when editing. Tbh, I didn't even know this was legal in java until you changed a few of them.

File: src/com/tictacgo/data/Piece.java (hunk 9 above)
Comment (apmcleod) [inline src/com/tictacgo/data/Piece.java:212]: Yes, I originally branched before those changes were made I think. Those should get changed automatically on merge.

File: src/com/tictacgo/TicTacGoGame.java (hunk 7 above)
Comment (apmcleod) [inline src/com/tictacgo/TicTacGoGame.java]: Do we want to do that for sure? The changes I've been doing on the direction-picker branch make this much shorter.

File: src/com/tictacgo/data/Piece.java (hunk 8 above)
Comment (apmcleod) [inline src/com/tictacgo/data/Piece.java:148]: Yeah, me too. I'll add as a new issue. See #25

File: src/com/tictacgo/TicTacGoGame.java (hunk 6 above)
Comment (apmcleod) [inline src/com/tictacgo/TicTacGoGame.java]: And an additional note, we don't actually need to save the entire Board. Really, all we need is a list of Pieces. Everything else can be generated on load. (See #26)

**Comment by apmcleod** [comment]
@hopskipnfall if you go commit by commit for this PR (look at commits from Jan 5 on), the exact changes are more clear. Realizing our dates might be different... you want the ones since "Manifest updated and new java file" inclusive.

File: src/com/tictacgo/TicTacGoMenuActivity.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+import android.os.Bundle;
+import android.view.View;
+import android.view.View.OnClickListener;
+import android.widget.RadioGroup;
+import android.widget.TextView;
+
+import com.tictacgo.data.Board.Player;
+
+
+public class TicTacGoMenuActivity extends Activity {
```
Comment (jimmyfnugent) [inline src/com/tictacgo/TicTacGoMenuActivity.java:17]: Javadoc

File: src/com/tictacgo/TicTacGoMenuActivity.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+  public static final String p1NameKey = "com.tictacgo.p1Name";
```
Comment (jimmyfnugent) [inline src/com/tictacgo/TicTacGoMenuActivity.java:19]: UPPERCASE_NAMES

Comment (apmcleod) [inline src/com/tictacgo/TicTacGoMenuActivity.java:19]: I changed this in #27 after your comment there.

File: src/com/tictacgo/TicTacGoMenuActivity.java (hunk 10 above)
Comment (apmcleod) [inline src/com/tictacgo/TicTacGoMenuActivity.java:17]: Added to all classes here.

File: src/com/tictacgo/TicTacGoGameActivity.java [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+
+  /**
+   * The FrameLayout in which the game is to be drawn
+   */
+  private FrameLayout fl;
+
+  /**
+   * OnClickListener to use for each piece
+   */
+  private View.OnClickListener pieceClicked;
```
Comment (jimmyfnugent) [inline src/com/tictacgo/TicTacGoGameActivity.java]: Let's make this onPieceClicked and onDirectionClicked.

File: src/com/tictacgo/TicTacGoGameActivity.java [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+          case Gravity.BOTTOM | Gravity.RIGHT:
+            xOffset = height / 2;
+            yOffset = height / 2;
+            break;
+          default:
+            xOffset = height / 4;
+            yOffset = height / 4;
+        }
+        directions.showAtLocation(fl, Gravity.NO_GRAVITY, xOffset, yOffset); //Make popup visible
+        int id = R.drawable.pieceodirection; //Which picture to use for directions (O)
```
Comment (jimmyfnugent) [inline src/com/tictacgo/TicTacGoGameActivity.java:141]: Not sure if this has been taken care of in another PR already, but resource files should be named in snake style (piece_o_direction). That can be in another PR though.

**Comment by hopskipnfall** [comment]
I think the reason why all of these old commits are being grouped into this PR is because you are making multiple PRs at the same time.  It might be easier if you wait until one PR is closed before filing the next.

**Comment by hopskipnfall** [comment]
LGTM

File: src/com/tictacgo/TicTacGoGameActivity.java (hunk 13 above)
Comment (apmcleod) [inline src/com/tictacgo/TicTacGoGameActivity.java:141]: Yeah. Added #29 since this seems to be coming up a lot.

File: src/com/tictacgo/TicTacGoGameActivity.java (hunk 12 above)
Comment (apmcleod) [inline src/com/tictacgo/TicTacGoGameActivity.java]: Changed here. (and also in direction-picker)

**Comment by jimmyfnugent** [comment]
LGTM
